File: benchmarks/c/channel_pipeline.c
#include <pthread.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int64_t *values;
    int capacity;
    int head;
    int tail;
    int count;
    int sealed;
    pthread_mutex_t mutex;
    pthread_cond_t not_empty;
    pthread_cond_t not_full;
} Queue;

typedef struct {
    int items;
    Queue *output;
} ProducerArgs;

typedef struct {
    Queue *input;
    Queue *output;
    int rounds;
} WorkerArgs;

static int env_int(const char *name, int fallback) {
    const char *raw = getenv(name);
    char *end = NULL;
    long value;
    if (raw == NULL || raw[0] == '\0') {
        return fallback;
    }
    value = strtol(raw, &end, 10);
    if (end == raw || *end != '\0' || value <= 0 || value > 2147483647L) {
        return fallback;
    }
    return (int)value;
}

static int64_t heavy_work(int64_t seed, int rounds) {
    int64_t acc = seed + 1;
    int i;
    for (i = 0; i < rounds; i++) {
        acc = (acc * 1103515245LL + 12345LL + i + seed) % 2147483647LL;
    }
    return acc % 1000003LL;
}

static int queue_init(Queue *queue, int capacity) {
    queue->values = (int64_t *)calloc((size_t)capacity, sizeof(int64_t));
    if (queue->values == NULL) {
        return 0;
    }
    queue->capacity = capacity;
    queue->head = 0;
    queue->tail = 0;
    queue->count = 0;
    queue->sealed = 0;
    pthread_mutex_init(&queue->mutex, NULL);
    pthread_cond_init(&queue->not_empty, NULL);
    pthread_cond_init(&queue->not_full, NULL);
    return 1;
}

static void queue_destroy(Queue *queue) {
    pthread_cond_destroy(&queue->not_full);
    pthread_cond_destroy(&queue->not_empty);
    pthread_mutex_destroy(&queue->mutex);
    free(queue->values);
}

static void queue_seal(Queue *queue) {
    pthread_mutex_lock(&queue->mutex);
    queue->sealed = 1;
    pthread_cond_broadcast(&queue->not_empty);
    pthread_cond_broadcast(&queue->not_full);
    pthread_mutex_unlock(&queue->mutex);
}

static int queue_send(Queue *queue, int64_t value) {
    pthread_mutex_lock(&queue->mutex);
    while (!queue->sealed && queue->count == queue->capacity) {
        pthread_cond_wait(&queue->not_full, &queue->mutex);
    }
    if (queue->sealed) {
        pthread_mutex_unlock(&queue->mutex);
        return 0;
    }
    queue->values[queue->tail] = value;
    queue->tail = (queue->tail + 1) % queue->capacity;
    queue->count++;
    pthread_cond_signal(&queue->not_empty);
    pthread_mutex_unlock(&queue->mutex);
    return 1;
}

static int queue_recv(Queue *queue, int64_t *value) {
    pthread_mutex_lock(&queue->mutex);
    while (!queue->sealed && queue->count == 0) {
        pthread_cond_wait(&queue->not_empty, &queue->mutex);
    }
    if (queue->count == 0 && queue->sealed) {
        pthread_mutex_unlock(&queue->mutex);
        return 0;
    }
    *value = queue->values[queue->head];
    queue->head = (queue->head + 1) % queue->capacity;
    queue->count--;
    pthread_cond_signal(&queue->not_full);
    pthread_mutex_unlock(&queue->mutex);
    return 1;
}

static void *producer_thread(void *ptr) {
    ProducerArgs *args = (ProducerArgs *)ptr;
    int i;
    for (i = 0; i < args->items; i++) {
        queue_send(args->output, i);
    }
    queue_seal(args->output);
    return NULL;
}

static void *worker_thread(void *ptr) {
    WorkerArgs *args = (WorkerArgs *)ptr;
    int64_t value;
    while (queue_recv(args->input, &value)) {
        queue_send(args->output, heavy_work(value, args->rounds));
    }
    return NULL;
}

int main(void) {
    int workers = env_int("BENCH_THREADS", 4);
    int items = env_int("BENCH_ITEMS", 20000);
    int rounds = env_int("BENCH_ROUNDS", 64);
    int capacity = workers * 4;
    Queue input;
    Queue output;
    ProducerArgs producer_args;
    WorkerArgs *worker_args = NULL;
    pthread_t producer;
    pthread_t *threads = NULL;
    int64_t checksum = 0;
    int received = 0;
    int i;

    if (!queue_init(&input, capacity)) {
        return 1;
    }
    if (!queue_init(&output, capacity)) {
        queue_destroy(&input);
        return 1;
    }

    threads = (pthread_t *)calloc((size_t)workers, sizeof(pthread_t));
    worker_args = (WorkerArgs *)calloc((size_t)workers, sizeof(WorkerArgs));
    if (threads == NULL || worker_args == NULL) {
        free(threads);
        free(worker_args);
        queue_destroy(&output);
        queue_destroy(&input);
        return 1;
    }

    producer_args.items = items;
    producer_args.output = &input;
    if (pthread_create(&producer, NULL, producer_thread, &producer_args) != 0) {
        free(threads);
        free(worker_args);
        queue_destroy(&output);
        queue_destroy(&input);
        return 1;
    }

    for (i = 0; i < workers; i++) {
        worker_args[i].input = &input;
        worker_args[i].output = &output;
        worker_args[i].rounds = rounds;
        if (pthread_create(&threads[i], NULL, worker_thread, &worker_args[i]) != 0) {
            return 1;
        }
    }

    while (received < items) {
        int64_t value;
        if (queue_recv(&output, &value)) {
            checksum += value;
            received++;
        }
    }

    pthread_join(producer, NULL);
    for (i = 0; i < workers; i++) {
        pthread_join(threads[i], NULL);
    }
    queue_seal(&output);

    printf("checksum: %lld\n", (long long)checksum);
    printf("processed: %d\n", received);
    free(threads);
    free(worker_args);
    queue_destroy(&output);
    queue_destroy(&input);
    return 0;
}

File: benchmarks/python/channel_pipeline.py
import os
import queue
import sys
import sysconfig
import threading


def require_free_threaded():
    if sys.version_info < (3, 14):
        raise SystemExit("Python concurrency benchmarks require Python 3.14+")
    if str(sysconfig.get_config_var("Py_GIL_DISABLED")) != "1":
        raise SystemExit("Python concurrency benchmarks require a free-threaded build")
    is_gil_enabled = getattr(sys, "_is_gil_enabled", None)
    if callable(is_gil_enabled) and is_gil_enabled():
        raise SystemExit("Python concurrency benchmarks require the GIL disabled")


def env_int(name, fallback):
    try:
        value = int(os.environ.get(name, ""))
    except ValueError:
        return fallback
    return value if value > 0 else fallback


def heavy_work(seed, rounds):
    acc = seed + 1
    for i in range(rounds):
        acc = (acc * 1103515245 + 12345 + i + seed) % 2147483647
    return acc % 1000003


def producer(items, input_queue, sentinel, workers):
    for i in range(items):
        input_queue.put(i)
    for _ in range(workers):
        input_queue.put(sentinel)


def worker(input_queue, output_queue, rounds, sentinel):
    while True:
        value = input_queue.get()
        if value is sentinel:
            return
        output_queue.put(heavy_work(value, rounds))


def main():
    require_free_threaded()
    workers = env_int("BENCH_THREADS", 4)
    items = env_int("BENCH_ITEMS", 20000)
    rounds = env_int("BENCH_ROUNDS", 64)
    capacity = workers * 4
    sentinel = object()
    input_queue = queue.Queue(maxsize=capacity)
    output_queue = queue.Queue(maxsize=capacity)

    producer_thread = threading.Thread(
        target=producer,
        args=(items, input_queue, sentinel, workers),
    )
    worker_threads = [
        threading.Thread(target=worker, args=(input_queue, output_queue, rounds, sentinel))
        for _ in range(workers)
    ]

    producer_thread.start()
    for thread in worker_threads:
        thread.start()

    checksum = 0
    for _ in range(items):
        checksum += output_queue.get()

    producer_thread.join()
    for thread in worker_threads:
        thread.join()

    print(f"checksum: {checksum}")
    print(f"processed: {items}")


if __name__ == "__main__":
    main()
